### PR TITLE
Export stateful executor states to`ThunderModule`

### LINKED ISSUE #2438: Connect TEv2 states to ThunderModule
## 🚀 Feature

Make TEv2 states accessible as attributes to the thunder module. This would make it more transparent to the user and help with testing/verifying TEv2's accuracy.

### PR DESCRIPTION
**Pull request by mattteochen**
## What does this PR do?

Closes #2438.

### Summary
- Added ExportStatefulExecutorsTransform (singleton) with a registry of export callbacks.
- Executors can register a callback to export runtime state post-execution.
- Currently integrates with `transformer_engine_ex`.

### Usage
The export is controlled by passing the transform to the transforms list:
```python
thunder.jit(model, executors=[...], transforms=[TransformerEngineTransform(), ExportStatefulExecutorsTransform()])
```
We might or might not go towards a compiler flag to make this more intuitive for the user (hence handling this transform automatically).

Whole example:
```python
from thunder.dev_utils.export_stateful_ex_transform import ExportStatefulExecutorsTransform
import torch
import torch.nn as nn
import thunder
from thunder.executors.transformer_engineex import transformer_engine_ex, TransformerEngineTransform
from transformer_engine.common import recipe
import transformer_engine.pytorch as te
from pprint import pprint

torch.manual_seed(42)

# Device and data type
dtype = torch.bfloat16
device = "cuda"

# Inputs (3D input)
x = torch.randn(3, 768, 4096, device=device, dtype=dtype, requires_grad=True)

# TE recipe
fp8_recipe = recipe.DelayedScaling()
# fp8_recipe = recipe.MXFP8BlockScaling()

# Dummy model
class Module(nn.Module):
    def __init__(self):
        super().__init__()
        self.w1 = nn.Parameter(torch.randn(4096, 4096, device=device, dtype=dtype))
        self.w2 = nn.Parameter(torch.randn(4096, 4096, device=device, dtype=dtype))
        self.w3 = nn.Parameter(torch.randn(4096, 4096, device=device, dtype=dtype))

    def forward(self, x):
        return torch.nn.functional.linear(torch.nn.functional.linear(torch.nn.functional.linear(x, self.w2), self.w3), self.w1)

model = Module()
jmodel = thunder.jit(model, executors=[transformer_engine_ex], transforms=[TransformerEngineTransform(), ExportStatefulExecutorsTransform()])

# Enable autocasting for the forward pass
for _ in range(2):
    with te.fp8_autocast(fp8_recipe=fp8_recipe):
        y = jmodel(torch.randn(3, 768, 4096, device=device, dtype=dtype, requires_grad=True))
        rep = jmodel.te_fp8_states()
        pprint(rep) # Visualize states information after operation
    grad_output = torch.randn_like(y)
    y.backward(grad_output)
    rep = jmodel.te_fp8_states(mode="backward")
    pprint(rep) # Visualize states information after operation
```

## Why it matters
- Not TE-specific: provides a reusable, uniform path for any stateful executor to export runtime state for debugging, validation, and reporting.

### CODE REVIEW

File: thunder/tests/test_export_stateful_ex_transform.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+    import transformer_engine as _te_mod  # noqa: F401
+    import transformer_engine.pytorch as te  # type: ignore
+    from transformer_engine.common import recipe  # type: ignore
+    from thunder.executors.transformer_engineex import transformer_engine_ex, TransformerEngineTransform
+
+    TE_AVAILABLE = True
+except Exception:
+    te = None  # type: ignore
+    recipe = None  # type: ignore
+    TE_AVAILABLE = False
```
Comment (crcrpar) [inline thunder/tests/test_export_stateful_ex_transform.py:26]: At glance I'm speculating that `pytest.importerskip("transformer_engine")` would be sufficing ([ref](https://docs.pytest.org/en/7.1.x/reference/reference.html#pytest-importorskip))

File: thunder/tests/test_export_stateful_ex_transform.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+    def thunder_model(x, fp8_recipe):
+        with te.fp8_autocast(fp8_recipe=fp8_recipe):
+            return jmodel(x)
+
+    train_model(thunder_model)
+
+    stats = jmodel.te_fp8_stats
+    from pprint import pprint
+
+    pprint(stats)
```
Comment (crcrpar) [inline thunder/tests/test_export_stateful_ex_transform.py]: is this a fallout of debugging?

File: thunder/tests/test_export_stateful_ex_transform.py (hunk 1 above)
Comment (mattteochen) [reply thunder/tests/test_export_stateful_ex_transform.py:26]: Yes, but I didn't want to skip the entire test file, as future tests might not depend on the `te` executor

File: thunder/dev_utils/export_stateful_ex_transform.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        cls._register_callbacks[name] = callback
+        cls._callback_attributes.append((name, instance, resolve_cb))
+
+    def transform_module(self, model) -> None:
+        assert model is not None
+        # Cache a weakref to the ThunderModule for later runtime export
+        self.tm_refs.append(weakref.ref(model))
+        # Initialize attributes on model
+        for name, instance, resolve_cb in self._callback_attributes:
+            setattr(model, name, instance(model, resolve_cb))
```
Comment (kshitij12345) [inline thunder/dev_utils/export_stateful_ex_transform.py:108]: I think this line leads to reference cycle -> model holds a reference to instance and instance holds a reference to the model.
